### ISSUE TITLE
`self` and `instance` types are converted to incorrect type in block argument

### ISSUE THREAD

**Issue by pocke** [soutaro/steep#804]
Since Steep 1.2.0, `self` and `instance` types are converted to an incorrect type if it is placed in a block argument.

For example:

```ruby
# test.rb
class C
  def block_self(&block)
    yield self
    block.call self
  end

  def block_instance(&block)
    yield self
    block.call self
  end

  def block_classname(&block)
    yield self
    block.call self
  end
end
```

```rbs
# test.rbs
class C
  def block_self: () { (self) -> void } -> void
  def block_instance: () { (instance) -> void } -> void
  def block_classname: () { (C) -> void } -> void
end
```

```ruby
# Steepfile
D = Steep::Diagnostic
target :lib do
  signature "."
  check "test.rb"
end
```

```console
$ bundle exec steep check
# Type checking files:

..................F.................................................................

test.rb:5:15: [error] Cannot pass a value of type `self` as an argument of type `^(self) -> void`
│   self <: ^(self) -> void
│     ::C <: ^(self) -> void
│
│ Diagnostic ID: Ruby::ArgumentTypeMismatch
│
└     block.call self
                 ~~~~

test.rb:10:15: [error] Cannot pass a value of type `self` as an argument of type `::Proc`
│   self <: ::Proc
│     ::C <: ::Proc
│       ::Object <: ::Proc
│         ::BasicObject <: ::Proc
│
│ Diagnostic ID: Ruby::ArgumentTypeMismatch
│
└     block.call self
                 ~~~~

Detected 2 problems from 1 file
```

I expected no errors for this file, but Steep displays two errors about `block.call self`.

I've confirmed that this behavior has been introduced since Steep v1.2.0. `steep check` displays these errors on v1.2.0 or later, but it doesn't display them on v1.1.1.

**Comment by ParadoxV5** [comment]
No errors with Steep 1.7.1, other than
```
[Steep 1.7.1] [typecheck:typecheck@1] [background] [#typecheck_source(path=ruby.rb)] [#type_check_file(ruby.rb@.)] [synthesize:(2:1)] [synthesize:(3:3)] [synthesize:(8:3)] [synthesize:(9:5)] [synthesize:(9:5)] [self <: instance] [::C <: instance] `T <: instance` doesn't hold generally, but testing it with `::C <: instance && instance <: ::C` for compatibility
```
* Possibly fixed by #1165

**Comment by mikegw** [comment]
Hi, just to clarify, is there a reason we get the warning?

To simplify the example above, for the following `.rbs` file:
```rbs
# foo.rbs
class Foo
  def call: () { (instance) -> void } -> void
end
```
using explicit block syntax passes `steep check` with no warnings:
```ruby
# foo.rb
class Foo
  def call(&blk)
    blk.call(self)
  end
end
```
but changing the code to yield and running `steep check` prints the "`T <: instance` doesn't hold generally..." warning:
```ruby
# foo.rb
class Foo
  def call
    yield self
  end
end
```
Is it expected that yielding generate warnings while capturing the block and calling it explicitly does not?

Of course, this isn't a big deal. I was using Steep v1.7.1 and ruby 3.3.4 when I saw this behavior, and it also happens with Steep v1.8.0.dev.1 . I'd be happy to try and submit a PR if this warning is not expected, but I'm new to using Steep so I want to make sure I'm not misunderstanding the intended behavior before I wade into the code.

Here's the full warning message, in case it helps:
```
[Steep 1.8.0.dev.1] [typecheck:typecheck@10] [background] [#typecheck_source(path=lib/foo.rb)] [#type_check_file(lib/foo.rb@lib)] [synthesize:(3:1)] [synthesize:(4:3)] [synthesize:(5:5)] [self <: instance] [::Foo <: instance] `T <: instance` doesn't hold generally, but testing it with `::Foo <: instance && instance <: ::Foo` for compatibility
```

By the way, thanks for all the work you do to make Steep the amazing tool that it is!
